### PR TITLE
Upper canvas retina scaling

### PR DESCRIPTION
**Pull request by cabaret**
@asturur, as requested in #5931, I've moved the retina scaling code from #5746 to a separate branch. 

Currently, the pointer offset is incorrect due to the retina scaling. Is this something I should fix in `getPointer` (`canvas.class.js:1310`) or would you suggest another place? I'll do the work if you guide me a bit ;)

Thanks.

### CODE REVIEW

**Comment by cabaret** [comment]
In `canvas.class.js`, if I use the following code, everything seems to work. Not sure if this is the correct approach though:

```
      if (boundsWidth === 0 || boundsHeight === 0) {
        // If bounds are not available (i.e. not visible), do not apply scale.
        cssScale = { width: 1, height: 1 };
      }
      else {
        cssScale = {
          width: upperCanvasEl.width / boundsWidth,
          height: upperCanvasEl.height / boundsHeight
        };
      }
      if (upperRetina) {
        cssScale = {
          width: cssScale.width / 2,
          height: cssScale.height / 2,
        };
      }
```

**Comment by asturur** [comment]
Going to look at it, i have a working example somewhere, i ll dig it up. I prefer to keep the one i have, so i'm sure it works already.

**Comment by asturur** [comment]
I have to introduce some breaking changes soon.
Skip the configuration option here, if retina scaling is enabled in general it will be for both canvases.

No need for extra config.

**Comment by asturur** [comment]
So this is my personal implementation in an app

this is the getPointer method of canvas.

```
  getPointer(e, ignoreZoom, upperCanvasEl) {
    const p = this.callSuper('getPointer', e, ignoreZoom, upperCanvasEl);
    const retinaScaling = this.getRetinaScaling();
    if (retinaScaling !== 1) {
      p.x /= retinaScaling;
      p.y /= retinaScaling;
    }
    return p;
  },
```

those lines can be added in the original function directly.

File: dist/fabric.js [hunk 1]
```diff
@@ -6872,10 +6873,17 @@ fabric.ElementsParser = function(elements, callback, options, reviver, parsingOp
       if (!this._isRetinaScaling()) {
         return;
       }
-      this.lowerCanvasEl.setAttribute('width', this.width * fabric.devicePixelRatio);
-      this.lowerCanvasEl.setAttribute('height', this.height * fabric.devicePixelRatio);
+      var scaleRatio = fabric.devicePixelRatio;
+      this.__initRetinaScaling(scaleRatio, this.lowerCanvasEl, this.contextContainer);
+      if (this.enableRetinaScalingUpper && this.upperCanvasEl) {
```
Comment (asturur) [inline dist/fabric.js]:
```suggestion
      if (this.upperCanvasEl) {
```

File: dist/fabric.js [hunk 2]
```diff
@@ -9650,6 +9658,18 @@ fabric.PatternBrush = fabric.util.createClass(fabric.PencilBrush, /** @lends fab
...
+     * Indicates if the upperCanvas is retina enhanced
+     * inserted to enable the feature but to not introduce a breaking change.
+     * In fabric 4.0 will be removed and the single enableRetinaScaling will influence
+     * both canvases.
+     * @type Boolean
+     * @since 3.2.0
+     * @deprecated, will be removed in 4.0
+     * @default false
+     */
+    enableRetinaScalingUpper: true,
```
Comment (asturur) [inline dist/fabric.js]: remove all together. Extra complication.

File: dist/fabric.js [hunk 3]
```diff
@@ -10669,22 +10689,24 @@ fabric.PatternBrush = fabric.util.createClass(fabric.PencilBrush, /** @lends fab
      * @throws {CANVAS_INIT_ERROR} If canvas can not be initialized
      */
     _createUpperCanvas: function () {
-      var lowerCanvasClass = this.lowerCanvasEl.className.replace(/\s*lower-canvas\s*/, '');
+      var lowerCanvasClass = this.lowerCanvasEl.className.replace(/\s*lower-canvas\s*/, ''),
+          lowerCanvasEl = this.lowerCanvasEl, upperCanvasEl = this.upperCanvasEl;
```
Comment (asturur) [inline dist/fabric.js]:
```suggestion
      var lowerCanvasEl = this.lowerCanvasEl, upperCanvasEl = this.upperCanvasEl;
          lowerCanvasClass = lowerCanvasEl.className.replace(/\s*lower-canvas\s*/, ''),
```

File: dist/fabric.js [hunk 4]
```diff
@@ -14602,6 +14626,7 @@ fabric.util.object.extend(fabric.StaticCanvas.prototype, /** @lends fabric.Stati
      * Sets "color" of an instance (alias of `set('fill', &hellip;)`)
      * @param {String} color Color value
      * @return {fabric.Object} thisArg
+     * @deprecated since 3.5.0
```
Comment (asturur) [inline dist/fabric.js]: i need to release a 3.5.0 just for those deprecations so that i can remove them in 4.0

File: src/canvas.class.js [hunk 5]
```diff
@@ -330,6 +330,18 @@
...
+     * Indicates if the upperCanvas is retina enhanced
+     * inserted to enable the feature but to not introduce a breaking change.
+     * In fabric 4.0 will be removed and the single enableRetinaScaling will influence
+     * both canvases.
+     * @type Boolean
+     * @since 3.2.0
+     * @deprecated, will be removed in 4.0
+     * @default false
+     */
+    enableRetinaScalingUpper: true,
```
Comment (asturur) [inline src/canvas.class.js]: Remove this all

File: src/canvas.class.js [hunk 6]
```diff
@@ -1309,7 +1321,7 @@
           bounds = upperCanvasEl.getBoundingClientRect(),
           boundsWidth = bounds.width || 0,
           boundsHeight = bounds.height || 0,
-          cssScale;
+          cssScale, upperRetina = this.enableRetinaScalingUpper;
```
Comment (asturur) [inline src/canvas.class.js]:
```suggestion
          cssScale;

File: src/static_canvas.class.js [hunk 7]
```diff
@@ -276,10 +276,17 @@
       if (!this._isRetinaScaling()) {
         return;
       }
-      this.lowerCanvasEl.setAttribute('width', this.width * fabric.devicePixelRatio);
-      this.lowerCanvasEl.setAttribute('height', this.height * fabric.devicePixelRatio);
+      var scaleRatio = fabric.devicePixelRatio;
+      this.__initRetinaScaling(scaleRatio, this.lowerCanvasEl, this.contextContainer);
+      if (this.enableRetinaScalingUpper && this.upperCanvasEl) {
```
Comment (asturur) [inline src/static_canvas.class.js]:
```suggestion
      if (this.upperCanvasEl) {
```

**Comment by cabaret** [comment]
Updated the PR, reverted previous dist mess and fixed your remarks.

**Comment by asturur** [comment]
wondering why this scaling is not already working correctly:

https://github.com/fabricjs/fabric.js/pull/5938/files#diff-9dc132ed9a0787568f773999ba577e2bR1341

Does it feels like working to you?
Can you try with also a container that is css stretched?

**Comment by cabaret** [comment]
Just tried this build in our project and it seems to work properly. 

Getting two failed tests apparently:

- number is halved because of retina scaling, I suppose? (57/2 ~= 28)
```
fabric.IText: hiddenTextarea does not move DOM
    ---
        actual: >
            28
        expected: >
            57
```

- not sure about this one:
```
 iText click interaction: click on editing itext make selection:changed fire
    ---
        actual: >
            1
        expected: >
            2
        stack: >
                at http://localhost:8080/3638/test/unit/itext_click_behaviour.js:145:14
        message: >
            Itext set the selectionStart
        Log: |
```

**Comment by cabaret** [comment]
@asturur I've started fixing the tests (differentiating between retina disabled and enabled) and they work fine locally but I can't seem to get them to work on the CI. Seems like it always returns the non-retina values. Any pointers? Thanks!

**Comment by cabaret** [comment]
Hey @asturur, did you have any time yet to take a look at this? Thanks!

**Comment by asturur** [comment]
Did not have time, i ll try today/tomorrow. Can you put me as contributor on your fork so i can push up changes?

**Comment by cabaret** [comment]
@asturur No problem. I've added you as a collaborator. Let me know where I can help!

**Comment by cabaret** [comment]
Hey @asturur, anything I can help with to move this forward? Thanks!

**Comment by asturur** [comment]
I'm sorry really under a pile of things to do.
Tomorrow i start an holiday, i have the notebook with me, i m sure i ll find time in the evenings to do something.

File: test/unit/itext.js [hunk 8]
```diff
@@ -1,735 +1,770 @@
 (function() {
   var canvas = this.canvas = new fabric.Canvas();
 
-  QUnit.module('fabric.IText', {
-    afterEach: function() {
+  QUnit.module('fabric.IText', function(hooks) {
+    hooks.afterEach(function() {
```
Comment (asturur) [inline test/unit/itext.js]: i m not expert with QUnit, can you explain me this change?

Comment (cabaret) [reply test/unit/itext.js]: I don't remember perfectly but I split up those tests in three blocks: independent of retina, retinaEnabled false and retinaEnabled true. I needed a way to specify the afterEach block for all tests and then the beforeEach calls for the nested modules. I found it this way in the QUnit docs but perhaps there are other ways. I think the reason I changed the line you commented on is to make the hooks calls the same throughout the file. Not quite sure anymore, sorry!

**Comment by asturur** [comment]
I just want to read the changes to tests because the diffs are so large i want to understand if is just indentation or there is more. I ll merge the next time i m on the computer.

**Comment by asturur** [comment]
image

Support for languages that may use composition is broken right now.
The hiddentextArea does not get positioned correctly.

**Comment by asturur** [comment]
gsplit2

Shadow in free drawing is broken

**Comment by asturur** [comment]
Things starts to be complicated. Retina scaling in Node is always 1, so some tests may be ineffective.
Anyway pencil brush is fixed and there is no way i can write a test that catch the regression for now.
We need to fix the textbox position.

**Comment by cabaret** [comment]
@asturur Can you give me some pointers on the problem with the textbox? I'm not sure I understand the issue. I'll gladly have a go at fixing it once I know what to look for ;)

**Comment by asturur** [comment]
in the function called `_calcTextareaPosition` there is some adjustment to do.
To be honest i did not expect it.
Probably we need to multiply or divide by the retina factor.

**Comment by cabaret** [comment]
What did you mean by `Support for languages that may use composition is broken right now.`? I'm not exactly sure what the hidden textbox is supposed to do or how to reproduce this. I'll take a look at the `_calcTextareaPosition` method.

**Comment by asturur** [comment]
you need to set your input language in japanese for example and type things.
You will notice that the overlay dropdown is displaced, like in my gif

**Comment by cabaret** [comment]
This is what I'm getting with Japanese: https://cloud.cupofco.de/qGuz2ZgB

**Comment by asturur** [comment]
that seems fine, but wasn't what i was getting at all.

**Comment by asturur** [comment]
this is what i get with latest branch running locally

image

**Comment by cabaret** [comment]
Hey @asturur, any idea on how I can reproduce this or next steps to get this merged eventually? Thanks.

**Comment by asturur** [comment]
ok let's merge!

**Pull request merged**